### ISSUE TITLE
Terminal freezes/jankiness when in input mode on 0.0.0-alpha+201502170110

### ISSUE THREAD

**Issue by telemachus** [neovim/neovim#2016]
After updating neovim yesterday and today on two separate machines, I'm getting sporadic freezes while in input mode. (The problem seems to be particularly related to rapidly hitting backspace to correct a typo and then entering the corrections, but I'm not 100% sure of that. It's sporadic.)

I know that this report is somewhat vague, and I apologize for that. Please let me know if I can supply any other particular information. I'm copying my version information below. (I'm on OSX 10.9.5 for both machines, and neovim was installed using Homebrew.) Thanks.

    NVIM 0.0.0-alpha+201502170110 (compiled Feb 18 2015 07:56:48)
    Commit: 66a03a7385dfd11503d22d8cc9dc6a972e27dcc2
    Build type: RelWithDebInfo
    Compilation: /usr/local/Library/ENV/4.3/clang -Wconversion -O2 -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -I/tmp/neovim-lB9irD/build/config -I/tmp/neovim-lB9irD/src -I/tmp/neovim-lB9irD/.deps/usr/include -I/tmp/neovim-lB9irD/.deps/usr/include -I/tmp/neovim-lB9irD/.deps/usr/include/luajit-2.0 -I/tmp/neovim-lB9irD/.deps/usr/include -I/usr/local/opt/gettext/include -I/usr/include -I/tmp/neovim-lB9irD/build/src/nvim/auto -I/tmp/neovim-lB9irD/build/include
    Compiled by _homebrew@ilium  Features included (+) or not (-):
    +acl             +farsi           -mouse_gpm       +tag_old_static
    +arabic          +file_in_path    +mouse_netterm   -tag_any_white
    +autocmd         +find_in_path    +mouse_sgr       +termresponse
    -browse          +folding         -mouse_sysmouse  +textobjects
    +byte_offset     -footer          +mouse_urxvt     +title
    +cindent         +fork()          +mouse_xterm     -toolbar
    -clipboard       +gettext         +multi_byte      +user_commands
    +cmdline_compl   +iconv           +multi_lang      +vertsplit
    +cmdline_hist    +insert_expand   +path_extra      +virtualedit
    +cmdline_info    +jumplist        +persistent_undo +visual
    +comments        +keymap          +postscript      +visualextra
    +conceal         +langmap         +printer         +viminfo
    +cscope          +libcall         +profile         +vreplace
    +cursorbind      +linebreak       +quickfix        +wildignore
    +cursorshape     +lispindent      +reltime         +wildmenu
    +dialog_con      +listcmds        +rightleft       +windows
    +digraphs        +localmap        +scrollbind      +writebackup
    -dnd             +menu            +signs           -X11
    -ebcdic          +mksession       +smartindent     -xfontset
    -emacs_tags      +modify_fname    +startuptime     -xsmp
    +eval            +mouse           +statusline      -xterm_clipboard
    +ex_extra        -mouseshape      +syntax          
    +extra_search    +mouse_dec       +tag_binary      
       system vimrc file: "$VIM/nvimrc"
         user vimrc file: "$HOME/.nvimrc"
     2nd user vimrc file: "~/.nvim/nvimrc"
          user exrc file: "$HOME/.exrc"
      fall-back for $VIM: "/usr/local/Cellar/neovim/HEAD/share/nvim"

**Comment by elmart** [comment]
@tarruda You know I'm already trying to narrow down the cause for this. But, as it seems to affect only OS X people (at least by now), just let me know if I can help you somehow.

**Comment by khalidchawtany** [comment]
I have this same issue. Just because of this, for now I have switched back to MacVim till this will be solved as this makes using NeoVim unbearable. :(

**Comment by kopischke** [comment]
I can confirm the random keypress processing lags on OS X 10.10.2, running NVIM 0.0.0-alpha+201502170110 (compiled Feb 17 2015 16:17:11), commit 66a03a7385dfd11503d22d8cc9dc6a972e27dcc2, built via Homebrew.

**Comment by tarruda** [comment]
Is everyone experiencing this lag also using YouCompleteMe?

**Comment by kopischke** [comment]
@tarruda nope, no YCM here.

**Comment by khalidchawtany** [comment]
Yep, I use YCM.

**Comment by tarruda** [comment]
> Yep, I use YCM.

Does the issue disappears if you disable YCM?

**Comment by elmart** [comment]
@tarruda I think the YCM thing was a red herring. I pointed to that because of the msgpack error line line in the log, but we later saw that was unrelated and happening always on exit. Also, @kopischke has the problem without YCM.

**Comment by khalidchawtany** [comment]
No, Disabling YCM does not solve the issue
@tarruda I find very strange UI updates. Let me explain. For example, I type something, the first two or three characters are shown and then the third and fourth are not shown till I switch to normal mode or I type another character.

Another problem which I think is relevant
In insert mode if I write some suff and then press ESC and type another character It does not switch to normal mode but types strange characters. e with two dots on it or e with '^' on it. Sorry I don't know what they are called.
![screen shot 2015-02-18 at 18 44 26](https://cloud.githubusercontent.com/assets/1681606/6250355/4cb35cb2-b79e-11e4-9f88-eecd95aa68bf.png)

**Comment by khalidchawtany** [comment]
If I press ESC and then wait a little then press ESC it switches to normal mode

**Comment by telemachus** [comment]
I don't use YCM. I do use Ultisnips, but I haven't had it crash on me yet. @khalidchawtany Have you run `pip install --update neovim` lately? I wonder if that's related to the crashes. Just a thought.

Here are my plugins, in case there are any other possible interaction questions:

    gist-vim
    sweater
    ultisnips
    vim-bats
    vim-lastplace
    vim-markdown
    vim-repeat
    vim-sol
    vim-surround
    vim-varia # Personal tweaks and a handful of colorschemes
    webapi-vim

**Comment by kopischke** [comment]
I can confirm both problems @khalidchawtany reports, without ever having used or installed YCM.

**Comment by khalidchawtany** [comment]
@telemachus, Yes, I have done ```pip install --update neovi```

**Comment by tarruda** [comment]
Ok, I will merge #2007 today which contain some easy fixes and then will focus on this problem

**Comment by tarruda** [comment]
Can someone test if the pygtk UI also has the same issue? I would be great to help me find the problem

**Comment by khalidchawtany** [comment]
Where to get **pygtk UI**? Is it OS X compatible? 
Sorry for asking, but I have no idea about pygtk.

**Comment by tarruda** [comment]
> Where to get pygtk UI? Is it OS X compatible? 
Sorry for asking, but I have no idea about pygtk.

My fault for not providing more details. Its a Gtk UI prototype I've written in python to test the new UI protocol and is installed with the python client(pip install --upgrade neovim)

Theoretically it should be compatible with OSX since its written in pure python/gtk+, but since I have no access to OSX I can't be sure. You need gtk3/gobject introspection + python bindings. I have written a script to install the required dependencies in a pyenv installation https://gist.github.com/tarruda/b67472898a7e4bd81572, though you may need to adapt it for OSX.

If you manage to run this UI in OSX without getting this lag then I will be sure this is a problem specific to the TUI that was merged a few days ago.

**Comment by tarruda** [comment]
@kopischke / @khalidchawtany Is the crash still present on latest master(after #2007 was merged)?

**Comment by kopischke** [comment]
@tarruda just rebuilt NVIM 0.0.0-alpha+201502181331 (compiled Feb 18 2015 18:05:58)                                                           
Commit: 366662d932551e558d10f09887ddf144ed5db34b – all issues persist. On the screenshot, the “é” appears where I pressed `ESC`:

![neovim input issues](https://cloud.githubusercontent.com/assets/80906/6252271/46a95f1a-b799-11e4-8caa-5f6c0e0c53b5.png)

I also have seen transient spurious control characters appear on the command line (`[^` if memory serves, among others).

**Comment by khalidchawtany** [comment]
Compiling .......

**Comment by telemachus** [comment]
@tarruda On OSX, under homebrew, it looks like pygobject3 and py2cairo are available--and homebrew would automatically pull in the other dependencies. Almost. It also requires XQuartz, which is a separate install. I won't be able to try this today, but might be able later this week.

Just to be clear, once these dependencies are installed, we would just need to reinstall python's neovim module? (Also how would the py/GTK gui be opened or started? I'm guessing it wouldn't be built as a standard OSX GUI app in /Applications. Is there a way to call it from the command line, à la `gvim`?)

**Comment by khalidchawtany** [comment]
@telemachus  you can start it using one of there
 ```pynvim```
or
```python -m neovim.ui.cli```

**Comment by tarruda** [comment]
@khalidchawtany you also need the -g flag for the gtk UI: `python -m neovim.ui.cli -g` or `pynvim -g` if the package is installed

**Comment by khalidchawtany** [comment]
Thanks @tarruda. What about the missing tickit issue. I cannot start it with the -g flag or without it.

**Comment by tarruda** [comment]
@khalidchawtany are you getting a missing tickit error when trying to start the UI with -g?

**Comment by khalidchawtany** [comment]
@tarruda Yes, I do get an error.
![screen shot 2015-02-18 at 21 08 39](https://cloud.githubusercontent.com/assets/1681606/6253436/6173a49a-b7b2-11e4-8304-6d08b89ddf78.png)

**Comment by tarruda** [comment]
@khalidchawtany that error is because you need to install pycairo

**Comment by khalidchawtany** [comment]
@tarruda that error is new. You are right. However, it was complaining about **tickitUI** before. I installed some packages that may have fixed it.

**Comment by tarruda** [comment]
> @tarruda that error is new. You are right. However, it was complaining about tickitUI before. I installed some packages that may have fixed it.

let me explain: When you run `python -m neovim.ui.cli` you are trying to start a terminal UI based on libtickit(it will fail if you dont have it installed). If you add the -g flag it will run the pygtk UI, which requires a different set of dependencies(gtk, cairo, gobject-introspection and all python bindings).

**Comment by renstrom** [comment]
I've been trying to gather some information about this. And here's a couple of things I've experienced:

- When using `nvim -u NONE` everything runs super smooth.
- Removing [sheerun/vim-polyglot](https://github.com/sheerun/vim-polyglot) (a language pack basically) seems to lower the freeze time (and perhaps the frequency as well?).
- Setting `syntax off` also makes it run a bit smoother.

This leads me to believe it's somewhat related to rendering colors/syntax highlighting...? Does that make sense at all? Or am I just seeing the effects of not having to bother with colors?

**Comment by bugeats** [comment]
Might be related to #2017.  Here's my notes on a a little reverting I did: https://github.com/neovim/neovim/issues/2017#issuecomment-74931374

**Comment by tarruda** [comment]
Any OSX user managed to test if this issue happens with the pygtk UI?

**Comment by tarruda** [comment]
Also, does it help if the python host is disabled?(eg: pip uninstall neovim)

**Comment by kopischke** [comment]
> Any OSX user managed to test if this issue happens with the pygtk UI?

Sorry, no, no time to try this ATM, but

> Also, does it help if the python host is disabled?(eg: pip uninstall neovim)

In a quick test run, it seems to alleviate the lagginess, but it does **not** help with `ESC` inserting spurious characters and only working as `Escape` after a delay.

**Comment by tarruda** [comment]
> In a quick test run, it seems to alleviate the lagginess, but it does not help with ESC inserting spurious characters and only working as Escape after a delay.

Ok thanks. Does `set ttimeoutlen=-1` helps with the slowness or the spurious characters?

**Comment by kopischke** [comment]
@tarruda I have updated my comment – the notice about alleviated lagginess was premature, sorry.

**Comment by kopischke** [comment]
> Does set ttimeoutlen=-1 helps with the slowness or the spurious characters?

That restores normal `ESC` behaviour, but the “burst mode” lagginess is unchanged (see my two previous comments).

**Comment by renstrom** [comment]
>Also, does it help if the python host is disabled?(eg: pip uninstall neovim)

No change. Just typing `:h ttimeoutlen<Enter>` takes like four seconds before it appears.

**Comment by tarruda** [comment]
I'm installing a mountain lion VM to debug this

**Comment by renstrom** [comment]
Also, I have no idea if https://github.com/neovim/neovim/issues/2024 is in anyway related. But I thought I should mention it.

**Comment by khalidchawtany** [comment]
@tarruda I managed to get **pygtk UI** (after a lot of suffering). I can confirm that these abnormal characters don't appear when I press ESC followed by another character.

**Comment by tarruda** [comment]
> @tarruda I managed to get pygtk UI (after a lot of suffering). I can confirm that these abnormal characters don't appear when I press ESC followed by another character.

@khalidchawtany does the pygtk UI suffers from any of the issues described here?

**Comment by khalidchawtany** [comment]
No, I dno't think so I just installed it so I am not sure if other problems may exist.

**Comment by khalidchawtany** [comment]
***NOT Talking about pygtk UI***
@tarruda  setting ```set ttimeoutlen=-1``` does help a bit if I press ESC and another letter. However, if I press ESC followed by a letter very fast then I still get strange characters.

**Comment by tarruda** [comment]
> @tarruda setting set ttimeoutlen=-1 does help a bit if I press ESC and another letter. However, if I press ESC followed by a letter very fast then I still get strange characters.

thanks, I'm currently configuring a OSX VM to debug this myself. Could you give me some feedback on the pygtk UI? If there are no issues with it then I will be sure that the problem lies in the tui module

**Comment by khalidchawtany** [comment]
@tarruda I had to install **cairo** with the **--with-x11** flag, If I compiled **cairo** using **--without-x11** flag, **pynvim -g** was reporting that ```Symbol not found: _cairo_xlib_surface_get_depth```.

As a result of running **pynvim** inside **X11**, I don't know which kind of problems I should report.

I see that there are some rendering issues. For example I have ```set cursorline``` an when I naviage throught the code I see that some traces of the cursor line is left. As I said before I don't know whether this is related to **X11** or **pynvim**.
![screen_shot_2015-02-19_at_16_14_48](https://cloud.githubusercontent.com/assets/1681606/6267163/c9a819ce-b852-11e4-91cc-0904bc4d1e5b.png)

**Comment by khalidchawtany** [comment]
Another thing is that when I type the cursor is moving away from the actual letter that is being typed.

![error](https://cloud.githubusercontent.com/assets/1681606/6267283/a820ed16-b853-11e4-8621-f144944f3da6.png)

**Comment by tarruda** [comment]
@khalidchawtany Thanks for the detailed report. These are drawing issues that must be fixed in the pygtk UI, what about lagging or esc inserting spurious characters, are you experiencing anything like that in the pygtk UI?

**Comment by khalidchawtany** [comment]
@tarruda  no, I don't have those problems in pygtk UI. :)

**Comment by elmart** [comment]
Regarding freezes, has anybody found a reliable way to make them happen?

**Comment by tarruda** [comment]
@elmart I forgot to update this thread, sorry. I've installed a OSX vm and may have a landed a fix in #2025(need more OSX users to test)

**Comment by kopischke** [comment]
@tarruda will test tomorrow and report back (middle of the night ATM this side of the Atlantic).

**Comment by telemachus** [comment]
@tarruda I've installed latest and so far, so good. I'll report back after working a full day tomorrow. Thanks.

**Comment by tarruda** [comment]
> @tarruda will test tomorrow and report back (middle of the night ATM this side of the Atlantic).

Thanks, looking forward to your feedback but I'm almost sure its fixed. The problem was that the new TUI replaced libuv pipe handle by the tty handle to read from stdin, and the tty handle seems to be broken on OSX

**Comment by kopischke** [comment]
I just re-installed via Homebrew to HEAD of @tarruda’s *tui-fixes-2* branch

```
NVIM 0.0.0-alpha+201502192025 (compiled Feb 20 2015 11:02:37)                                                           
Commit: da364c4f156bb47f2f263af68e30d86528c75252
```

First findings:
- input lagginess / burst mode freezing seems to be gone (not tested over a prolonged period yet – *caveat empteor* – but looking good so far)
- insert mode delayed `ESC` response / spurious characters insert is unchanged. Setting `ttimeoutlen=-1` alleviates the issue somewhat, but not entirely.

@tarruda would you prefer transferring discussion of the latter to #2017?

**Comment by tarruda** [comment]
> @tarruda would you prefer transferring discussion of the latter to #2017?

Since thats a separate issue, yes

**Comment by kopischke** [comment]
@tarruda then it looks as if it is time for a “Closes #2016” commit :). Great work, thank you!

**Comment by elmart** [comment]
I've never been able to reproduce this reliably, so I can't be sure it's fixed. But I've been working the whole morning with this without problems. So, it seems fixed, or at least, much better than before.
Thus, I think #2025 should be merged asap, to avoid more people on OSX picking the wrong version.
Great job, @tarruda :+1:

**Comment by adambiggs** [comment]
So far so good over here too :+1: 

Thanks for the quick fix!
